File: 백준/Gold/1405. 미친 로봇/미친 로봇.py
import sys
input = sys.stdin.readline

N, e, w, n, s = map(int,input().split())


if N==1 or e*w==n*s==0:
    print(1.0)
    sys.exit()

size = 2*N+1

visited =[[False]*(size) for i in range(size)]

result = 0
probability = 1

cur_pos = (0,0)
visited[0][0] = True

direction =[]
for p, (dx,dy) in zip([e,w,n,s],[(0,1),(0,-1),(1,0),(-1,0)]):
    if p:
        direction.append([p,dx,dy])
def dfs(depth):
    global cur_pos,N,result, probability,size
    # print('cur',cur_pos)
    if depth == N:
        result += probability
        # print(probability,result)
        # print('visited',*visited,sep='\n')
        return
    x,y = cur_pos
    for p, dx, dy in direction:
        nx,ny = (dx+x)%size, (dy+y)%size
        # print('dx,dy',[dx,dy],'nx,ny',[nx,ny],visited[nx][ny])
        if visited[nx][ny]:
            continue
        probability*=p
        visited[nx][ny]=True
        cur_pos = (nx,ny)
        dfs(depth+1)
        probability//=p
        cur_pos = (x,y)
        visited[nx][ny]=False
        
dfs(0)

print(result/100**N)


    

File: 백준/Gold/1405. 미친 로봇/미친 로봇.java
import java.io.*;
import java.util.*;

public class Main {
    static int N, size;
    static double result = 0.0;
    static double probability = 1.0;
    static boolean[][] visited;
    static int curX = 0, curY = 0;

    static class Dir {
        int p, dx, dy;
        Dir(int p, int dx, int dy) {
            this.p = p; this.dx = dx; this.dy = dy;
        }
    }

    static List<Dir> direction = new ArrayList<>();

    static void dfs(int depth) {
        if (depth == N) {
            result += probability;
            return;
        }
        int x = curX, y = curY;
        for (Dir d : direction) {
            int nx = (d.dx + x + size) % size;
            int ny = (d.dy + y + size) % size;
            if (visited[nx][ny]) continue;

            probability *= d.p;
            visited[nx][ny] = true;
            curX = nx; curY = ny;

            dfs(depth + 1);

            probability /= d.p;
            curX = x; curY = y;
            visited[nx][ny] = false;
        }
    }

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        N = Integer.parseInt(st.nextToken());
        int e = Integer.parseInt(st.nextToken());
        int w = Integer.parseInt(st.nextToken());
        int n = Integer.parseInt(st.nextToken());
        int s = Integer.parseInt(st.nextToken());

        if (N == 1 || (e * w == 0 && n * s == 0)) {
            System.out.println(1.0);
            return;
        }

        size = 2 * N + 1;
        visited = new boolean[size][size];
        visited[0][0] = true;

        if (e > 0) direction.add(new Dir(e, 0, 1));
        if (w > 0) direction.add(new Dir(w, 0, -1));
        if (n > 0) direction.add(new Dir(n, 1, 0));
        if (s > 0) direction.add(new Dir(s, -1, 0));

        dfs(0);

        double denom = Math.pow(100, N);
        System.out.println(result / denom);
    }
}
